spectator-ext-stackdriver/src/main/java/com/netflix/spectator/stackdriver/StackdriverWriter.java

```diff
@@ -0,0 +1,760 @@
+   */
+  private final String applicationName;
+
+  /**
+   * A unique id that distinguishes our data from other replicas with the same
+   * applicationName (in the same project).
+   */
+  private String instanceId;
+
+  /**
+   * Filters Spectator metrics down to those we want to log with Stackdriver.
+   */
+  private final MeasurementFilter measurementFilter;
+
+  /**
+   * HACK: Stackdriver descriptors are immutable descriptors requiring
+   *       labels be known (and declared) up front.
+   *  Some types we are sharing across services that use different labels.
+   *  This is a mechanism allowing external injection of these special cases.
+   *  The hints are only used when the need to create a descriptor arises.
+   *  Descriptors are only created once in the lifetime of a project (unless
+   *  the label is explicitly deleted, erasing all history, at some point).
+   *
+   *  We could externally create the labels, but doing so dynamically feels
+   *  more maintainable.
+   *
+   *  The key in this table is the desired Stackdriver descriptor name.
+   *
+   *  Note that this is not thread safe. Presumably we write to the hints
+   *  only on startup, and from within one thread so this doesnt matter.
+   */
+  private final Map<String, List<String>> labelHints
+      = new HashMap<String, List<String>>();
+
+  /**
+   * If Stackdriver were to return an error complaining about an TimeSeries
+   * data element we are trying to write, it references that by its index
+   * into the array we sent. That is pretty much useless to debug since we
+   * have no idea what or how the data is organized.
+   *
+   * This method is intended to interpret the Stackdriver error message
+   * in the context of the request we made and return the referenced element
+   * so that it can be logged in more explicit detail to understand what
+   * happend (most likely what label was missing from what descriptor type).
+   */
+  static public TimeSeries
+  findProblematicTimeSeriesElement(String msg, List<TimeSeries> nextN) {
+    String regex = "timeSeries\\[(\\d+?)\\]\\.metric\\.labels\\[(\\d+?)\\]";
+    Matcher matcher = Pattern.compile(regex).matcher(msg);
+    if (matcher.find()) {
+      int tsIndex = Integer.parseInt(matcher.group(1)) - 1;
+      TimeSeries ts = nextN.get(tsIndex);
+        return ts;
+    }
+    return null;
+  }
+
+  /**
+   * Factory method to create a writer.
+   *
+   * The customTypeNamespace is intended to act as a namespace for the
+   * stackdriver metrics found, such as the system name (e.g. spinnaker).
+   * This will help deconflict our metrics with another system that is
+   * also using Stackdriver Custom Metrics in the same project as us and
+   * might want to have CustomTypeDescriptors with the same name.
+   */
+  public static StackdriverWriter makeStackdriverWriter(
+      String customTypeNamespace,
+      String projectName,
+      String applicationName,
```

Comment:
Not sure I understand what the difference between these three are from a user perspective.